### ISSUE TITLE
Why the model performance decrease a lot after fine-tuning?

### ISSUE THREAD

**Issue by yxliao95** [shtoshni/fast-coref#13]
Hi @shtoshni,

I am trying to fine-tune this model to a custom dataset as I mentioned in issue #11. 

Before doing so, I used the datasets you provided (litbank and quizbowl) to simulate this process. However, the (conll) F1 score decreased a lot after I fine-tuned the model. Do you have any idea why this happen?  

Here are the experiment details:

[joint_best] = longformer_coreference_joint encoder + joint_best model
[ontonotes_best] = longformer_coreference_notonotes encoder + ontonotes_best model
[Pretrained encoder + initial model] = pretrained longformer_coreference_joint encoder + initial model

We didn't fine-tune the doc encoder in any experiments (config.model.doc_encoder.finetune=false).
All the configs remain almost the same except for the ones we presented in the hyperparameters column.

As you can see in No.1, I used the joint_best model and evaluated it on the litbank (split No.0) datset. The conll F1 score was exactly what we were expecting. 
In No.2, I used only 1 litbank record to fine-tune the model. The final F1 score was still good. I think this might prove that the code is working fine. (However, the F1 score during training decreased a lot, yet I don't know why.)
<img width="681" alt="截屏2022-08-02 13 58 37" src="https://user-images.githubusercontent.com/20765016/182380455-afc0a08d-8610-4d4d-be01-c42fbc60d563.png">

Then, in No.3, I finetuned the joint_best model on the litbank_0 dataset with different settings, and the final performance decreased a lot. I made a few assumptions about this decrease, like overfitting (as I was using litbank to fine-tune the joint_best model), and wrong gradient update direction (as I used config: "eval_per_k_steps=10").
<img width="678" alt="截屏2022-08-02 11 26 00" src="https://user-images.githubusercontent.com/20765016/182353093-bf48a246-b3a9-49d1-b76f-957fc83c4de9.png">

For the assumption of the wrong gradient update direction, I increased the eval_per_k_steps to 180 and 400. But the results remained almost the same (see No.4 and No.5).
<img width="681" alt="截屏2022-08-02 11 27 07" src="https://user-images.githubusercontent.com/20765016/182353293-b0144802-fa79-444c-b22f-6bb83cee1654.png">

Considering the litbank dataset was used to train the joint_best model, I tried to fine-tune the model with a different dataset to avoid overfitting (see No.6-11). I used the quizbowl dataset and split the 400 test records into train/dev/test for fine-tuning, like the litbank dataset. This time, the result improved after fine-tuning (No.6 vs No.9). But the F1 scores were quite low.
<img width="695" alt="截屏2022-08-01 14 43 30" src="https://user-images.githubusercontent.com/20765016/182161392-adb82135-a7b6-4109-946b-3fe9dbd33a87.png">

I also tried to train the model from scratch (see No.12-13). In these two experiments, we were using the pre-trained longformer_coreference_joint as an encoder and didn't fine-tune it. The results were very close to No.3-5.
<img width="684" alt="截屏2022-08-02 11 17 30" src="https://user-images.githubusercontent.com/20765016/182351531-5a8d2d9c-e303-4182-90bc-618547330d72.png">

In short, there are two questions:

1. I fine-tuned the pre-trained model on litbank. The model performance was much lower than your best model. I was wondering why the performance gap happened. Is there anything I missed or misconfigured?

2. And I was also wondering why the results computed by the evaluation metrics you implemented in your program differ from the CoNLL metrics, for example:
```
[2022-08-01 22:39:40,706][HYDRA] Test
[2022-08-01 22:39:40,706][HYDRA] Dataset: LitBank
[2022-08-01 22:39:40,706][HYDRA] Dataset: litbank, Cluster Threshold: 1
[2022-08-01 22:39:40,706][HYDRA] Evaluating on 10 examples
[2022-08-01 22:40:57,327][HYDRA] F-score: 74.0 , MUC: 86.6, Bcub: 74.7, CEAFE: 60.8
Using CoNLL scorer
...
[2022-08-01 22:40:58,602][HYDRA] (CoNLL) F-score : 61.8, MUC: 86.6, Bcub: 68.1, CEAFE: 30.6
...
[2022-08-01 22:40:58,602][HYDRA] Oracle F-score: 0.875
```

**Comment by shtoshni** [comment]
@liaoooyx: I'm busy till this month-end. Will respond to you next month.

**Comment by pitrack** [comment]
(I'm just an observer subscribed to this repo and commenting somewhat randomly -- I didn't carefully read the full post but I have some guesses/suggestions/things to check.)

> 2. And I was also wondering why the results computed by the evaluation metrics you implemented in your program differ from the CoNLL metrics, for example:

The metrics might differ depending on whether the metric is evaluating singletons. I'm not sure which CoNLL scorer you are using, in some implementations of them, singleton clusters might get completely ignored. This is perfectly fine for CoNLL 2012 Shared task since OntoNotes does not contain any, and in fact, some models predict singletons *because they know they will get pruned/ignored by the scorer.*

On the other hand, some datasets like litbank and QBCoref have singletons and they need to be evaluated, i.e. the model can't cheat and overpredict singletons (that will hurt score precision) or output no singletons at all (that would hurt recall). Furthermore, the evaluation metric is not allowed to ignore singletons (that would give the wrong result).

So in short, I'd also recommend checking for qbcoref and litbank that 

1) your model output contains singletons
2) the evaluation script is evaluating them.

--

For the other question. A couple of casual comments: litbank evaluation is done with k(=10) folds, so the score on one of the folds would not be the same as the full dataset score that you sometimes see reported. For checking if the metrics/model config is compatible and set up correctly, you can always test on the training set and score should reach 100.

**Comment by yxliao95** [comment]
Thanks a lot for your advice! It really helps! I will check them later.
